jgrapht-core/src/main/java/org/jgrapht/alg/matching/blossom/v5/KolmogorovMinimumWeightPerfectMatching.java

```diff
@@ -0,0 +1,889 @@
+     * Constructs a new instance of the algorithm with the specified {@code options}
+     *
+     * @param graph   the graph a minimum weight perfect matching would be searched in
+     * @param options the options which define the strategies for the initialization and dual updates
+     */
+    public KolmogorovMinimumWeightPerfectMatching(Graph<V, E> graph, BlossomVOptions options) {
+        Objects.requireNonNull(graph);
+        if ((graph.vertexSet().size() & 1) == 1) {
+            throw new IllegalArgumentException(NO_PERFECT_MATCHING);
+        } else if (graph.getType().isDirected()) {
+            this.graph = new AsUndirectedGraph<>(graph);
+        } else {
+            this.graph = graph;
+        }
+        this.options = Objects.requireNonNull(options);
+    }
+
+    /**
+     * Computes and returns a minimum weight perfect matching in the {@code graph}. See the class description
+     * for the relative definitions and algorithm description.
+     *
+     * @return the perfect matching in the {@code graph} of minimum weight
+     */
+    @Override
+    public MatchingAlgorithm.Matching<V, E> getMatching() {
+        if (matching == null) {
+            lazyComputeMinimumWeightPerfectMatching();
+        }
+        return matching;
+    }
+
+    /**
+     * Returns the computed solution to the dual linear program with respect to the
+     * minimum weight perfect matching linear program formulation.
+     *
+     * @return the solution to the dual linear program formulated on the {@code graph}
+     */
+    public DualSolution getDualSolution() {
+        dualSolution = lazyComputeDualSolution();
+        return dualSolution;
+    }
+
+    /**
+     * Performs an optimality test after the perfect matching is computed. More precisely,
+     * checks whether dual variables of all pseudonodes and resulting slacks of all edges are non-negative and
+     * that slacks of all matched edges are exactly 0. Since the algorithm uses floating point arithmetic,
+     * this check is done with precision of {@link KolmogorovMinimumWeightPerfectMatching#EPS}
+     *
+     * @return true iff the assigned dual variable satisfy the dual linear program formulation and
+     * complementary slackness conditions are also satisfied. The total error must not exceed EPS
+     */
+    public boolean testOptimality() {
+        lazyComputeMinimumWeightPerfectMatching();
+        return getError() < EPS; // getError() won't return -1 since matching != null
+    }
+
+    /**
+     * Computes the error in the solution to the dual linear program. More precisely, the total error
+     * equals to the sum of:
+     * <ul>
+     * <li>Absolute value of edge slack if it negative or the edge is matched</li>
+     * <li>Absolute value of pseudonode variable if it is negative</li>
+     * </ul>
+     *
+     * @return the total numeric error
+     */
+    public double getError() {
+        lazyComputeMinimumWeightPerfectMatching();
+        BlossomVEdge edge;
```

Comment:
In general, variables should be declared at the point of first use (the only exception is variables which need to be scoped outside of a loop, or if you are coding in K&R C).  This comment applies throughout this entire PR.